[Catalogs] [Delta Lake] Add initial support for reading from Delta Lake.

clarkzinzow opened the pull request:
This PR adds initial support for reading from Delta Lake. This is done using the Python bindings for `delta-rs`, i.e. the `deltalake` package.

Closes #1865 
Closes #1878 

## TODOs

- [x] Add support for reading partitioned tables (requires virtual column support)
- [x] Add support for partition pruning via filter pushdown on (virtual) partition columns
- [x] Add stats to scan tasks
- [ ] (Future PR) Add support for deletion vectors
- [ ] (Future PR) Add support for column mappings

samster25 commented on daft/delta_lake/delta_lake_scan.py:
```diff
@@ -0,0 +1,187 @@
+
+    def to_scan_tasks(self, pushdowns: Pushdowns) -> Iterator[ScanTask]:
+        # TODO(Clark): Push limit and filter expressions into deltalake action fetch, to prune the files returned.
+        paths = self._table.file_uris()
```
Is there a reason we can't use the paths from `add_actions` and have to use `self._table.file_uris()`?

clarkzinzow replied:
Yep, the paths in `add_actions` only consists of the file name, rather than the full path/URI.

samster25 replied:
cool, got it

samster25 commented on daft/delta_lake/delta_lake_scan.py:
```diff
@@ -0,0 +1,187 @@
+            file_format_config = FileFormatConfig.from_parquet_config(ParquetSourceConfig())
+
+            if "partition_values" in row and len(row["partition_values"]) > 0:
+                partition_values = daft.table.Table.from_pydict(
```
You want to encode the dtype of the partition column field when creating the series here otherwise you may have a lossy cast due to type inference. See how we do it for iceberg
https://github.com/Eventual-Inc/Daft/blob/main/daft/iceberg/iceberg_scan.py#L109

samster25 commented on daft/delta_lake/delta_lake_scan.py:
```diff
@@ -0,0 +1,187 @@
+            )
+        # NOTE: We don't assume that table.file_uris() and table.get_add_actions() return file entries in the same order, just to be safe.
+        # TODO(Clark): Keep action table in pyarrow form for the sake of efficiency and dtype handling?
+        actions: dict[str, dict] = {row["path"]: row for row in add_actions.to_pylist()}
```
It actually may be more efficient to do our partition pruning by converting the `add_actions` table to a Daft Table and applying our predicate there. We should do that in a future add on.

clarkzinzow replied:
Since we no longer need the `.file_uris()` fetch in addition to the add actions fetch, I went back to operating directly on the pyarrow `RecordBatch`. When I add virtual column + filter pushdown support in the next PR, I'll probably convert `add_actions` to a Daft table so we can directly apply our filter expressions to the partition values in the table!

samster25 commented on tests/integration/delta_lake/test_table_read.py:
```diff
@@ -0,0 +1,76 @@
+import pandas as pd
+import pytest
+
+deltalake = pytest.importorskip("deltalake")
```
you need to add `deltalake` to our requirements-dev and to our pyproject.toml (pip install getdaft[deltalake]) otherwise this test will never be run.

samster25 commented on tests/integration/delta_lake/test_table_read.py:
```diff
@@ -0,0 +1,76 @@
+    for part_idx in range(request.param):
+        part_df = base_df.copy()
+        part_df["part_idx"] = pd.Series([part_idx] * 3)
+        deltalake.write_deltalake(path, part_df, mode="append")
```
you should add some partition columns otherwise the predicate pushdown wont really do much

samster25 replied:
got it

samster25 commented on tests/integration/delta_lake/test_table_read.py:
```diff
@@ -0,0 +1,76 @@
+    # log.
+    yield path, list(reversed(dfs))
+
+
```
We can also probably copy all the iceberg tables into deltalake so we can test those as well!

samster25 commented on src/daft-plan/src/optimization/rules/push_down_filter.rs:
```diff
@@ -116,26 +115,55 @@ impl OptimizerRule for PushDownFilter {
+                            // 1. Identity transform partition filters, which can then be dropped from the data-level filter.
+                            // 2. Non-identity transform partition filters, which must still be applied on the data.
+                            // 3. Filter referencing both partition and data columns, which must be applied to the data.
+                            // 4. Data filters (no partition column references), which will be applied on the data.
```
can we combine 2 and 4? It doesn't look like we use them individually anywhere

samster25 commented on src/daft-plan/src/optimization/rules/push_down_filter.rs:
```diff
@@ -116,26 +115,55 @@ impl OptimizerRule for PushDownFilter {
                         }
                         let new_predicate = external_info.pushdowns().filters.as_ref().map(|f| predicate.and(f)).unwrap_or(predicate.clone());
-                        let partition_filter = if let ExternalInfo::Scan(ScanExternalInfo {scan_op,  ..}) = &external_info {
+                        let PredicateGroups { partition_filter, non_identity_partition_filter_for_data, partition_and_data_filter, data_only_filter } = if let ExternalInfo::Scan(ScanExternalInfo {scan_op,  ..}) = &external_info {
```
given that we now bucket things into these categories and add a filter manually, we can now split the input expression, filtering out the ones that `has_udf`.

samster25 commented on src/daft-plan/src/optimization/rules/push_down_filter.rs:
```diff
@@ -116,26 +115,55 @@ impl OptimizerRule for PushDownFilter {
-                        let partition_filter = if let ExternalInfo::Scan(ScanExternalInfo {scan_op,  ..}) = &external_info {
+                        let PredicateGroups { partition_filter, non_identity_partition_filter_for_data, partition_and_data_filter, data_only_filter } = if let ExternalInfo::Scan(ScanExternalInfo {scan_op,  ..}) = &external_info {
+                            // We split the predicate into four groups:
+                            // 1. Identity transform partition filters, which can then be dropped from the data-level filter.
```
This subtle change here is that before we would only translate expressions to partition filters for allowlisted operations (<, <=, ==, >=, >, is_null, etc) but now it will allow all predicate expressions through. I think there may be some implications for this so we should think this through

samster25 commented on src/daft-scan/src/expr_rewriter.rs:
```diff
@@ -49,12 +49,84 @@ fn apply_partitioning_expr(expr: Expr, pfield: &PartitionField) -> Option<Expr>
+            if let Expr::Column(col_name) = e {
+                if let Some(pfield) = pfields_map.get(col_name.as_ref()) {
+                    all_data_keys = false;
+                    if !matches!(pfield.transform, Some(PartitionTransform::Identity) | None) {
```
as discussed in prior comment, this will allow all predicate expressions though rather than only the ones through allowed operations

clarkzinzow commented:
@jaychia once this is deemed good to merge, we'll need to override the required CI steps again, since this PR removes the (py runner, pyarrow==6.0.1, Ubuntu, Python 3.7) entry from the matrix.

samster25 commented on .github/workflows/python-package.yml:
```diff
@@ -23,12 +23,7 @@ jobs:
-        - daft-runner: py
-          python-version: '3.10'
-          pyarrow-version: 6.0.1
+        - pyarrow-version: 6.0.1
```
as discussed, we should still have ubuntu + pyarrow 6.0.1 in the matrix.

samster25 commented on pyproject.toml:
```diff
@@ -5,7 +5,8 @@ requires = ["maturin>=1.2.0,<1.3.0"]
 authors = [{name = "Eventual Inc", email = "daft@eventualcomputing.com"}]
 dependencies = [
-  "pyarrow >= 6.0.1",
+  "pyarrow >= 12.0.0; platform_system != 'Windows'",
```
we should leave the pin here at 6.0.1, we have some users are using pyarrow 7.0.0 or pyarrow 8.0.0

samster25 reviewed (approved):
Looks great! lets just enable pyarrow 6.0.1 in ubuntu but skip the deltalake test suites when that's the case.

clarkzinzow commented:
@samster25  I added back the pyarrow >= 6.0.1 lower bound and added it back to the CI matrix; I added an extra CI step to override the `deltalake` version if pyarrow==6.0.1 is installed, hopefully that works!

Also added S3 test coverage via a local mock S3 server via `moto` - I managed to get a session-level fixture working, so all deltalake tests will run against both plain local files and S3 files, and data generation fixtures are shared.
